### PR TITLE
[10.0] Supplier invoice and Cash discount

### PR DESCRIPTION
**Pull request by benwillig**
Theses modules add the concept of cash discount if an invoice has been paid entirely before a due date.

- ```account_cash_discount_base```: Compute a discount on an invoice based on a rate
- ```account_cash_discount_payment```: Handle the cash discount in a payment order
- ```account_cash_discount_writeoff```: Create a writeoff when confirming the payment order to mark the invoice as 'paid'

### CODE REVIEW

**Comment by benwillig** [comment]
cc @chrisdec @sbidoul

File: account_cash_discount_base/README.rst
Comment (sbidoul) [inline account_cash_discount_base/README.rst:22]: This module was written to allow you to define cash discount (aka early payment discount)
on customer or supplier invoices. 

On customer invoices, you can add a discount delay and
a discount percentage. At invoice validation, the discount due date is generated
from the delay. The amount with discount deducted is computed at the same time
as taxes. This module also adds cash discount informations on customer invoice
report.

On supplier invoices, you can set the discount percentage. For the discount due
date, you can fill it directly or set a delay. In this case, discount due date
is still computed at invoice's validation.

You can decide to calculate the discount amount based on the total price taxes
included or not.

File: account_cash_discount_base/README.rst
Comment (sbidoul) [inline account_cash_discount_base/README.rst]: Remove this section

File: account_cash_discount_base/models/account_invoice.py
Comment (sbidoul) [inline account_cash_discount_base/models/account_invoice.py:86]: a good old if then else is easier to read

File: account_cash_discount_base/models/account_invoice.py
Comment (sbidoul) [inline account_cash_discount_base/models/account_invoice.py]: why 0.0 + ... ?

File: account_cash_discount_base/README.rst
Comment (sbidoul) [inline account_cash_discount_base/README.rst]: This sounds dangerous. I think it's better to freeze these amounts when validating the invoice. As the readme says: `The amount with discount deducted is computed at the same time as taxes`.

File: account_cash_discount_base/README.rst
Comment (sbidoul) [inline account_cash_discount_base/README.rst]: Remove or put a real screenshot.

File: account_cash_discount_base/models/account_invoice.py
Comment (sbidoul) [inline account_cash_discount_base/models/account_invoice.py:188]: Why is this test necessary?

File: account_cash_discount_base/models/account_invoice.py
Comment (sbidoul) [inline account_cash_discount_base/models/account_invoice.py:227]: inv.id is not really helpful for the user (use inv.number instead, at this one it must have one?)

File: account_cash_discount_payment/README.rst
Comment (sbidoul) [inline account_cash_discount_payment/README.rst]: remove this section

File: account_cash_discount_payment/README.rst
Comment (sbidoul) [inline account_cash_discount_payment/README.rst:15]: Extend Account Cash Discount Base and Payment Order modules to let you
add invoices on payment order considering the cash discount on it.
When an invoice with cash discount is added to a payment order,
he amount proposed on payment line will be the residual amount discount deducted,
provided that the due date has not expired.

This module also offers to search by discount due date to add invoices on
payment order and add discount amount on payment line.

File: account_cash_discount_payment/models/account_payment_line.py
Comment (sbidoul) [inline account_cash_discount_payment/models/account_payment_line.py:30]: Can't we do the economy of this field? I think we can rely on the invoice total and enable the cash discount payment only if the invoice residual is equal to the invoice total. Ie if there is no invoice, disable pay with discount, and if the discount is partially paid, disable pay with discount. @chrisdec

File: account_cash_discount_payment/wizards/account_payment_line_create.py
Comment (sbidoul) [inline account_cash_discount_payment/wizards/account_payment_line_create.py]: Do we need start + end? For due date, the test is `('date_maturity', '<=', self.due_date)`, I think we can do the same for discount due date, it's easier and it's the most common use case.

File: account_cash_discount_write_off/README.rst
Comment (sbidoul) [inline account_cash_discount_write_off/README.rst:11]: Better have an independent flag to control tax adjustments.

File: account_cash_discount_write_off/models/account_payment_order.py
Comment (sbidoul) [inline account_cash_discount_write_off/models/account_payment_order.py]: prefix these methods with `_`

File: account_cash_discount_write_off/models/account_payment_order.py
Comment (sbidoul) [inline account_cash_discount_write_off/models/account_payment_order.py]: Post depending on payment mode config: https://github.com/OCA/bank-payment/blob/e752ebe943eb410f97fd419945b5a5fe45b0568a/account_payment_order/models/account_payment_order.py#L497-L498.

File: account_cash_discount_write_off/models/account_payment_line.py
Comment (sbidoul) [inline account_cash_discount_write_off/models/account_payment_line.py:40]: `not move_line.reconciled` instead of `has_partial_reconciliation`?

File: account_cash_discount_write_off/models/account_payment_line.py
Comment (sbidoul) [inline account_cash_discount_write_off/models/account_payment_line.py]: `_` prefix

File: account_cash_discount_write_off/models/account_payment_line.py
Comment (sbidoul) [inline account_cash_discount_write_off/models/account_payment_line.py]: not necessary: the partner on a move line is the accounting partner

File: account_cash_discount_write_off/models/account_payment_line.py
Comment (sbidoul) [inline account_cash_discount_write_off/models/account_payment_line.py]: use an independent flag for that

File: account_cash_discount_write_off/models/account_payment_line.py
Comment (sbidoul) [inline account_cash_discount_write_off/models/account_payment_line.py:60]: why this test?

File: account_cash_discount_write_off/models/account_payment_line.py
Comment (sbidoul) [reply account_cash_discount_write_off/models/account_payment_line.py:40]: perhaps we should do this writeoff only when the amound paid is exactly the cash discount amount on the invoice?

File: account_cash_discount_write_off/models/account_payment_line.py
Comment (sbidoul) [inline account_cash_discount_write_off/models/account_payment_line.py]: s/list()/[]/

File: account_cash_discount_write_off/models/account_payment_line.py
Comment (sbidoul) [inline account_cash_discount_write_off/models/account_payment_line.py]: Perhaps work on the move lines that have a tax_id field instead of relying on the invoice tax line?

**Review by sbidoul — COMMENTED** [review]
Looks very clean :) A few comments and questions.

File: account_cash_discount_write_off/README.rst
Comment (benwillig) [reply account_cash_discount_write_off/README.rst:11]: But would it be coherent to compute the discount based on the price (excluding taxes) and having the "Tax Adjustments" checked. What should we do in theses situations ?

File: account_cash_discount_base/models/account_invoice.py
Comment (benwillig) [reply account_cash_discount_base/models/account_invoice.py:188]: I should check the ```discount_percent``` instead of the ```discount_amount```. This test was to prevent to compute a discount due date if it was not needed.

But I'm wondering if we should add a boolean field which would tells if we pay the invoice using the discount or not. If this boolean equals to True, we could maybe force the user to fill the discount_percent and discount_delay.

File: account_cash_discount_base/models/account_invoice.py
Comment (benwillig) [reply account_cash_discount_base/models/account_invoice.py:227]: We can't use the number of the invoice because when the message is displayed, we rollback the transaction so we lose the invoice number.

File: account_cash_discount_payment/wizards/account_payment_line_create.py
Comment (benwillig) [reply account_cash_discount_payment/wizards/account_payment_line_create.py]: I don't know if start/end range is really needed, we should ask @chrisdec. But in my opinion, we could have a single date indeed.

File: account_cash_discount_write_off/models/account_payment_line.py
Comment (benwillig) [reply account_cash_discount_write_off/models/account_payment_line.py:60]: I agree that this method is called only when ```check_cash_discount_write_off_creation``` returns true. And this method returns True if the move line is linked to an invoice. So yes, this test might have no sense. but then I'm sure there won't be a huge stacktrace in the future because of a specific case.

File: account_cash_discount_write_off/models/account_payment_line.py
Comment (benwillig) [reply account_cash_discount_write_off/models/account_payment_line.py]: It would be more complicated I think. Because, based on the ```tax_ids``` field, we should get the corresponding move lines, and we could finally compute the amount.

File: account_cash_discount_base/models/account_invoice.py
Comment (benwillig) [reply account_cash_discount_base/models/account_invoice.py]: :alien: I think I was in an other dimension ^^

File: account_cash_discount_base/README.rst
Comment (benwillig) [reply account_cash_discount_base/README.rst]: Should we store the ```cash_discount_base_amount_type``` field on the invoice and populate it with an onchange? Or just set the discount fields as stored and remove the method dependency on ```company_id.cash_discount_base_amount_type```?

**Comment by benwillig** [comment]
I uptated my code based on your reviews and the discussions we had. To give a little summary of what I did:
- Invoice discount amounts are now stored
- The pay_with_discount flag on the payment line can only equals to True when the payment line is linked to an invoice which has a discount
- When changing the pay_with_discount flag, the amount is changed based on the invoice total and invoice discount
- A boolean flag has been added to know if we do tax adjustments. At the moment, it's a computed method which will simply assing True if the base amount type for the discount include all taxes. If it's change in the future, there will be less work to do
- The tax adjustments are now created based on the tax move lines
- Prefix some methods with a _
- Updated readmes

**Comment by sbidoul** [comment]
Runbot is orange due to a warning on an onchange.

**Comment by chrisdec** [comment]
Here are the results of first Functional tests:

1.	Could you please add a button on payment order lines to check or uncheck “Pay with discount”. It’s quite annoying to open the line each time you want to change the checkbox. 
2.	When you add journal items to payment orders, if day date < discount date,  the checkbox “Pay with discount” is not checked (Good) but the amount on payment order line is the amount with discount (Not good). 
3.	When you select move lines to put in your payment order, you don’t see cash discount amount. It could be great to add cash discount amount and discount due date in this screen (like in V8). 
4.	When you change “Bill date” on invoice, discount due date must be (re-)computed. 
5.	I can’t find a way to change VAT Base. The V8 module “account_invoice_manual_tax" is not available in V10?
6.	When you work with “Including all taxes”, you have to adapt base taxes and not only taxes on writeoff journal entry.

Thanks :)

**Comment by benwillig** [comment]
4. At the moment, we only use the invoice date to compute the discount due date. Should it be based only on the bill date?
5. No at the moment, this module is not going to be migrated.
6. So if we need to adapt base taxes, how would the invoice journal entry be?

File: account_cash_discount_write_off/models/account_payment_line.py
Comment (sbidoul) [inline account_cash_discount_write_off/models/account_payment_line.py]: @chrisdec @benwillig so here we should add the base taxes (tax_ids) in the writeoff? And therefore split the writeoff in several pieces for each base tax used?

Comment (chrisdec) [reply account_cash_discount_write_off/models/account_payment_line.py]: Yes. We have to split the writeoff in several lines but not in several pieces I think.

**Comment by benwillig** [comment]
@chrisdec I updated the module based on your review.

**Comment by chrisdec** [comment]
Hello, 

Here are the results of second Functional tests:

1. Why the button on payment order lines is "+" and "-" and not a checkbox? I think a checkbox would be simpler if it's possible

2. Point 3 of old tests is not good for me : Now, I see the discount amount and discount due date in the "Selected move lines to create Transactions" table (which is good :)) but I would like to see those fields in the "Add : move lines window" if it's possible. 

3. When you work with "Including tax" option, you don't use the write-off account defined on company (debit and credit = zero) but when you work with "Excluding tax" option, you use it. For me, it's not good. The write-off amount must be used in both option. Write-off account must replaced the expense account in the write-off journal entry. 

4. When you select move lines to add to payment order, if you choose to use "Discount due date", normally, the proposed move lines are all move lines with discount due date > selected date. Here, I also see move lines with discount due date < selected date when I click on "Add an item"

Thanks :)

**Comment by benwillig** [comment]
@chrisdec Everything should be fixed now. Could you check again?

**Comment by chrisdec** [comment]
Hello, 

Functional tests are ok for me. 

Thanks :)

**Comment by rafaelbn** [comment]
Interesting @carlosdauden @pedrobaeza

File: account_cash_discount_base/README.rst
Comment (rafaelbn) [inline account_cash_discount_base/README.rst:9]: @benwillig @sbidoul is this similar to https://github.com/OCA/account-payment/pull/184 ?

Comment (sbidoul) [reply account_cash_discount_base/README.rst:9]: Hi @rafaelbn ! This is the same domain, for sure. #184 seems to be about configuring the payment term only. Here we make it easy to pay supplier invoices with early payment discount and account_payment_order (including automatic reconciliation, proper handling of taxes with configurable rules etc). So there is little overlap it seems. At some point we considered using payment terms, which we rejected (I don't remember all the details now but one of the reason is that it was much more complex to handle)?

File: account_cash_discount_base/models/account_invoice.py
Comment (rousseldenis) [inline account_cash_discount_base/models/account_invoice.py:24]: @benwillig I would have added
```python
digits=dp.get_precision('Discount')
```

File: account_cash_discount_base/models/account_payment_term.py
Comment (rousseldenis) [inline account_cash_discount_base/models/account_payment_term.py]: @benwillig Same as above about dp

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review

**Review by rousseldenis — APPROVED** [review]

File: account_cash_discount_base/models/account_invoice.py
Comment (rousseldenis) [reply account_cash_discount_base/models/account_invoice.py:24]: Solved

**Comment by rousseldenis** [comment]
@sbidoul @rafaelbn Could you update your review as we could merge this?

**Review by rafaelbn — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Can you please squash commit history a bit?

**Comment by benwillig** [comment]
@pedrobaeza Still 18 commits but it's better than 54 I guess ^^

**Review by pedrobaeza — DISMISSED** [review]
What I don't get is the need of the extra module ` account_cash_discount_write_off` instead of an option at payment mode level.

**Comment by benwillig** [comment]
@pedrobaeza This could be an option indeed. But the code is more clear and it will be easier for the migration.

**Comment by rousseldenis** [comment]
I think we can move forward on this as it was proposed quite a year ago. Workin on several production environments.

**Comment by pedrobaeza** [comment]
@rousseldenis I agree to merge base module and the payment one, but not the third, so we can:

- Split these first 2 here, and merge, leaving the third.
- Merge already the 3rd one into the second in this PR, and merge all.

**Comment by rousseldenis** [comment]
@pedrobaeza IMHO there is no added value doing this. Things are well functionnally splitted.

**Comment by pedrobaeza** [comment]
The third module has no reason for being split, as there are no different dependencies and it's as simple as having the field in the parent one, and leaving empty means that there's no writeoff. Over-engineering things with lots of modules has not too much sense IMO.

**Comment by sbidoul** [comment]
@pedrobaeza IMO you are going too far in blocking this. 

The modularity does not harm at all. Ok, you prefer a slightly different design, but the current design has nothing wrong in itself. I would have been fine if you had brought this topic when we were creating the module, but now it's very late. 

The change you propose can be done later in another PR. There is good test coverage, it will be easy to refactor.

If we want to grow the community, we must remove roadblocks. And you must know these kind of blocking comments discourage even my colleagues at ACSONE from contributing to OCA. So imagine the effect it must have on less motivated or new contributors.

So please let's merge this: it's working well and has proven useful in at least 3 different countries.
I'm happy to leave it in Beta status until the redesign you propose has been addressed.

Finally, note we could perfectly have merged this before because we had enough approvals. We are just refraining from merging modules with only approvals from our own company.

**Comment by pedrobaeza** [comment]
If we merge, later then the argument would be that this can't be removed in a stable version. That's why I block it. Any way, I dismiss my review and do whatever you think better, but not with my approval.

**Comment by sbidoul** [comment]
Thanks @pedrobaeza 

We have two approvals from functional people from different companies, and two technical approvals from Acsone, and a remark on modularity by Pedro.

So let's merge in beta status.

**Pull request merged**
